Objects/perf_trampoline.c

```diff
@@ -0,0 +1,447 @@
+{
+    py_trampoline trampoline = (py_trampoline)code_arena->current_addr;
+    code_arena->size_left -= code_arena->code_size;
+    code_arena->current_addr += code_arena->code_size;
+    return trampoline;
+}
+
+static inline py_trampoline
+compile_trampoline(void)
+{
+    if ((code_arena == NULL) ||
+        (code_arena->size_left <= code_arena->code_size)) {
+        if (new_code_arena() < 0) {
+            return NULL;
+        }
+    }
+    assert(code_arena->size_left <= code_arena->size);
+    return code_arena_new_code(code_arena);
+}
+
+static PyObject *
+py_trampoline_evaluator(PyThreadState *ts, _PyInterpreterFrame *frame,
+                        int throw)
+{
+    if (perf_status == PERF_STATUS_FAILED ||
+        perf_status == PERF_STATUS_NO_INIT) {
+        goto default_eval;
+    }
+    PyCodeObject *co = frame->f_code;
+    py_trampoline f = NULL;
+    _PyCode_GetExtra((PyObject *)co, extra_code_index, (void **)&f);
+    if (f == NULL) {
+        // This is the first time we see this code object so we need
+        // to compile a trampoline for it.
+        if (extra_code_index == -1) {
+            extra_code_index = _PyEval_RequestCodeExtraIndex(NULL);
+        }
+        py_trampoline new_trampoline = compile_trampoline();
+        if (new_trampoline == NULL) {
+            goto default_eval;
+        }
+        trampoline_api.write_state(trampoline_api.state, new_trampoline,
+                                   code_arena->code_size, co);
+        _PyCode_SetExtra((PyObject *)co, extra_code_index,
+                         (void *)new_trampoline);
+        f = new_trampoline;
+    }
+    assert(f != NULL);
+    return f(_PyEval_EvalFrameDefault, ts, frame, throw);
+default_eval:
+    // Something failed, fall back to the default evaluator.
+    return _PyEval_EvalFrameDefault(ts, frame, throw);
+}
+#endif  // _PY_HAVE_PERF_TRAMPOLINE
+
+int
+_PyIsPerfTrampolineActive(void)
+{
+#ifdef _PY_HAVE_PERF_TRAMPOLINE
+    PyThreadState *tstate = _PyThreadState_GET();
+    return tstate->interp->eval_frame == py_trampoline_evaluator;
+#endif
+    return 0;
+}
+
+int
+_PyPerfTrampoline_SetCallbacks(trampoline_state_init init_state,
+                               trampoline_state_write write_state,
+                               trampoline_state_free free_state)
+{
+#ifdef _PY_HAVE_PERF_TRAMPOLINE
+    if (trampoline_api.state) {
+        Py_FatalError("Trampoline state already initialized");
+        return -1;
+    }
+    trampoline_api.init_state = init_state;
+    trampoline_api.write_state = write_state;
+    trampoline_api.free_state = free_state;
+    void *state = trampoline_api.init_state();
+    if (state == NULL) {
+        return -1;
+    }
+    trampoline_api.state = state;
+    perf_status = PERF_STATUS_OK;
+#endif
+    return 0;
+}
+
+int
+_PyPerfTrampoline_Init(int activate)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (!activate) {
+        tstate->interp->eval_frame = NULL;
+    }
```

Comment:
This will disable any other tool using the eval frame APIs, I suggest:

```suggestion
    if (!activate && tstate->interp->eval_frame == py_trampoline_evaluator ) {
        tstate->interp->eval_frame = NULL;
    }
```

Reply:
Other tools are incompatible with this one so not sure if is worth doing this.

Reply:
In that case we should consider raising an exception that a tool is already active rather than overriding it.

Reply:
OK makes sense will do the change.